src/lib/query-string.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * Encodes string by stricter rules
+ * see https://github.com/kevva/strict-uri-encode
+ *
+ * @param {String} string
+ */
+const encode = string => encodeURIComponent(string).replace(/[!'()*]/g, x => `%${x.charCodeAt(0).toString(16).toUpperCase()}`);
+
+/**
+ * Returns an array, where first element is the slice till separator and the second is the rest.
+ *
+ * @param {String} string
+ * @param {String} separator
+ */
+const splitOnFirst = (string, separator) => {
+    const separatorIndex = string.indexOf(separator);
+    if (separatorIndex === -1) {
+        return [string];
+    }
+    return [
+        string.slice(0, separatorIndex),
+        string.slice(separatorIndex + separator.length),
+    ];
+};
+
+/**
+ * Sorts array or object by keys
+ *
+ * @param {Array|Object} input
+ */
+const keysSorter = (input) => {
+    if (Array.isArray(input)) {
+        return input.sort();
+    }
+
+    if (typeof input === 'object') {
+        return keysSorter(Object.keys(input))
+            .sort((a, b) => Number(a) - Number(b))
+            .map(key => input[key]);
+    }
+
+    return input;
+};
+
+/**
+ * Formats the object by adding new keys
+ *
+ * @param {String} key
+ * @param {String} value
+ * @param {Object} accumulator
+ */
+const decoderFormatter = (key, value, accumulator) => {
+    if (accumulator[key] === undefined) {
+        accumulator[key] = value;
+        return;
+    }
+
+    accumulator[key] = [].concat(accumulator[key], value);
+};
```

Comment:
Let's aim to write pure functions. They are easier to test and less prone to bugs.
E.g.:
```
const decoderFormatter = (key, value, accumulator) => {
    if (accumulator[key] === undefined) {
        return accumulator[key] = value;
    }

    return [].concat(accumulator[key], value);
};
//...
ret[key] = decoderFormatter(decodeURIComponent(key), value, ret);
```

Also, it's more common to name function as verbs rather than nouns, e.g. _formatParsedEntry_, formatDecodedVal, normalizeDecodedVal, (your better version), etc. instead of _decoderFormatter_.